backend/test_observer/controllers/test_executions/end_test.py

```diff
@@ -0,0 +1,181 @@
+
+router = APIRouter()
+
+
+@router.put("/end-test")
+def end_test_execution(request: EndTestExecutionRequest, db: Session = Depends(get_db)):
+    test_execution = _find_related_test_execution(request, db)
+
+    if test_execution is None:
+        raise HTTPException(status_code=404, detail="Related TestExecution not found")
+
+    delete_previous_results(db, test_execution)
+    _store_test_results(db, request.test_results, test_execution)
+
+    has_failures = test_execution.has_failures
+
+    test_execution.status = (
+        TestExecutionStatus.FAILED if has_failures else TestExecutionStatus.PASSED
+    )
+
+    prev_test_execution = _get_previous_test_execution(db, test_execution)
+
+    if (
+        prev_test_execution
+        and not has_failures
+        and prev_test_execution.is_approved
+        and _ran_all_previously_run_cases(prev_test_execution, test_execution)
+    ):
+        test_execution.review_decision = [
+            TestExecutionReviewDecision.APPROVED_ALL_TESTS_PASS
+        ]
+
+    if request.c3_link is not None:
+        test_execution.c3_link = request.c3_link
+
+    db.commit()
+
+
+def _ran_all_previously_run_cases(
+    prev_test_execution: TestExecution, test_execution: TestExecution
+) -> bool:
+    prev_test_cases = {
+        tr.test_case.name
+        for tr in prev_test_execution.test_results
+        if tr.status != TestResultStatus.SKIPPED
+    }
+
+    test_cases = {
+        tr.test_case.name
+        for tr in test_execution.test_results
+        if tr.status != TestResultStatus.SKIPPED
+    }
+
+    return prev_test_cases.issubset(test_cases)
+
+
+def _get_previous_test_execution(
+    db: Session, test_execution: TestExecution
+) -> TestExecution | None:
+    artefact = test_execution.artefact_build.artefact
+    prev_artefact = _get_previous_artefact(db, artefact)
+
+    if prev_artefact is None:
+        return None
+
+    query = (
+        select(TestExecution)
+        .join(ArtefactBuild)
+        .where(
+            ArtefactBuild.artefact_id == prev_artefact.id,
+            TestExecution.environment_id == test_execution.environment_id,
+        )
+        .order_by(ArtefactBuild.revision.desc())
+        .limit(1)
+        .options(
+            joinedload(TestExecution.test_results).joinedload(TestResult.test_case)
+        )
+    )
+
+    return db.execute(query).unique().scalar_one_or_none()
+
+
+def _get_previous_artefact(db: Session, artefact: Artefact) -> Artefact | None:
+    family = artefact.stage.family
+
+    query = select(Artefact).where(
+        Artefact.name == artefact.name, Artefact.id != artefact.id
+    )
+
+    if family.name == FamilyName.SNAP.value:
+        query = query.where(Artefact.track == artefact.track)
+
+    if family.name == FamilyName.DEB.value:
+        query = query.where(
+            Artefact.series == artefact.series, Artefact.repo == artefact.repo
+        )
+
+    query = query.order_by(Artefact.id.desc()).limit(1)
+
```

Comment:
I think with our current setup we have NULL values for series/repo for snap artefacts, and similarly for deb artefacts we have NULL values for track fields. Maybe we can simplify this code by just trying to filter based on these three fields, not having separate logic for snaps/debs

Reply:
Good point